<a href="https://colab.research.google.com/github/stefanusbayuw/hacktiv8/blob/main/GC1_Test_Kurang_Exploreation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Graded Callange 1

## Perkenalan

Stefanus Bayu Waskito -- FTDS Batch Remote-18

 Pada data yang akan diiimpor ke colab merupakan data mengenai kasus Covid - 19 yang memiliki kolom berupa `province_state`, `country_region`, `date`, `latitude`, `longitude`, `location_geom`, `confirmed`, `deaths`, `recovered`, `active`, `fips`, `admin2` dan `combined_key`. Dari data yang ada ada beberapa kolom data yang dapat diubah namanya agar memudahkan dalam pengolahan data. Data yang dapat diubah yaitu :
 

*   `location_geom`. `location_geom` dapat diubah namanya menjadi `coordinates` karena nilai yang terdapat di setiap kolom berisi koordinat yang berupa penggabungan anatara longitude dan latitude.
*  ` admin2`. `admin2` dapat diubah menjadi `city_name` karena yang string berisi nama kota. 
*   `combined_key`. `combined key` berisi penggabungan isi cell dari kolom `province_state`, `country_region` dan `city_name` maka combined_key diubah namanya menjadi `location`.

Setelah melakukan perubahan nama kolom dengan tujuan mempermudah dalam pengolahan data, selanjutnya dilakukan manipulasi dengan menghapus kolom yang sekiranya tidak diperlukan dalam pengolahan data. Kolom yang akan dihapus yaitu:
*   `latitude` dihapus karena kolom tersebut merupakan kolom lintang yang mana nilai dari kolom tersebut sudah ada dalam `coordinates`
*   `longitude` dihapus karena kolom tersebut merupakan kolom bujur yang mana nilai dari kolom tersebut sudah ada dalam `coordinates`.
*   `fips` dihapus karena isi data `fips` adalah 5 digit kode wilayah yang hanya ditemukan di Amerika. Pada tabel, kode fips juga ditemukan hanya pada `country_region` yang bernilai/beisi 'US' selain itu tidak ditemukan kode fips.

Setelah proses penghapusan kolom yang tidak diperlukan dan melakukan perubahan nama kolom yang bertujuan mempermudah dalam pengolahan data, yang dilakukan selanjutnya adalah mengisi nilai null dan NaN pada kolom `city_name` dan `province_state`. Pada `city_name` nilai null dan NaN diisi dengan City sedangkan pada `province_state` diisi dengan Province. Mengapa nilai null dan NaN pada `city_name` dan `province_state` tidek dihapus saja? Mengisi nilai null dan NaN dilakukan karena setiap lokasi sudah memiliki koordinat tertentu sehingga kita dapat membedakan lokasi berdasarkan `coordinates`.

Query SQL untuk eksplorasi awal :
*   `SELECT * FROM bigquery-public-data.covid19_jhu_csse.summary /*Menmpilkan  preview data dari tabel Summary dengan*/`
*   `select count(distinct(province_state)) from bigquery-public-data.covid19_jhu_csse.summary /*Menampilkan jumlah data preview_state dari tabel summary yang menampilkan data sebanyak 759 data berupa list nama propinsi yang berarti ada 759 nama propinsi yang ada dalam kolom*/`
*   `select count(province_state) from bigquery-public-data.covid19_jhu_csse.summary /*Menampilkan  jumlah semua data preview_state dari tabel summary yang menampilkan data sebanyak 3.937.894 data berupa nama propinsi */`
*   `select count(distinct(country_region)) from bigquery-public-data.covid19_jhu_csse.summary /*Menampilkan  jumlah data coutry_region dari tabel summary yang output nya menampilkan total sebanyak 252 data berupa nama negara yang berarti ada 252 negara yang ada dalam kolom*/`
*   `select count(country_region) from bigquery-public-data.covid19_jhu_csse.summary /*Menampilkan  jumlah data preview_state dari tabel summary yang menampilkan data sebanyak 4.123.519 data berupa list nama propinsi */`
*   `select min(date) from bigquery-public-data.covid19_jhu_csse.summary/*Menampilkan data dengan nilai minimum, maksud minimum disini adalah tanggal paling awal di kolom data yaitu tanggal 2020-01-22*/`
*   `select max(date) from bigquery-public-data.covid19_jhu_csse.summary /*Menampilkan data dengan nilai maksimum, maksud mmaksimum disini adalah tanggal paling akhir (data yang diinput paling baru) di kolom data yaitu tanggal 2023-02-02*/`
*   `select count(distinct(date)) from bigquery-public-data.covid19_jhu_csse.summary where date between '2020-01-22' and '2023-02-02' /*mencari range dari antara data tanggal awal samapi tanggal akhir dimasukkan. Terdapat total 1108 hari dari tanggal awal sampai akhir*/`
*   `select sum(confirmed) from bigquery-public-data.covid19_jhu_csse.summary/*menghitung total keseluruhan total yang terkonfirmasi positif covid19 di tabel summary dengan asumsi data tercatat antara tanggal 2020-01-22 sampai dengan 2023-02-02 dengan total keseluruhan konfirmasi positif sebanyak 292.975.622.316*/`
*   `select sum(deaths) from bigquery-public-data.covid19_jhu_csse.summary/*menghitung total keseluruhan total yang meninggal akibat covid19 di tabel summary dengan asumsi data tercatat antara tanggal 2020-01-22 sampai dengan 2023-02-02 dengan total keseluruhan meninggal sebanyak 4.136.071.911*/`
*   `select round(avg(active)) from bigquery-public-data.covid19_jhu_csse.summary /*mengetahui rataan kasus aktif dari tabel summary yaitu sebanyak 6557 kasus. Penggunaan round() disini digunakan untuk membulatkan nilai hasil rataan*/`
*   `select count(country_region) from bigquery-public-data.covid19_jhu_csse.summary where fips is null /*menghitung jumlah data di kolom country_region yang fips-nya null. Total 739.155 kolom di country_region yang fips = null*/`
*   `select count(country_region) from bigquery-public-data.covid19_jhu_csse.summary where country_region = 'US' /*menghitung jumlah data di kolom country_region yang berisi string 'US'. Total 3.401.792 kolom di country_region yang berisi 'US'*/`
*   `select count(country_region) from bigquery-public-data.covid19_jhu_csse.summary where country_region != 'US' /*menghitung jumlah data di kolom country_region yang berisi string selain 'US'. Total 725.734 kolom di country_region yang berisi string selain 'US'*/`
*   `select count(country_region) from bigquery-public-data.covid19_jhu_csse.summary /*menghitung jumlah data di kolom country_region*/`


























## Import Library

In [1]:
import pandas as pd
import numpy as np

## Data Loading

Import Data "`[h8dsft_P0GC1_Stefanus-Bayu-Waskito.csv]`" dari Google Drive.

In [26]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [27]:
df=pd.read_csv('gdrive/My Drive/Hacktiv8/h8dsft_P0GC1_Stefanus-Bayu-Waskito.csv')

### Eksplorasi Data Sederhana

In [28]:
df.shape

(4127536, 13)

`df.shape` digunakan untuk menampilkan jumlah keseluruhan kolom dan baris data dari file .csv yang telah diimport ke jupyter notebook. Pada dataframe terdapat 4.127.536 baris data dan 13 kolom.

In [29]:
df.head(10)

province_state country_region        date   latitude  longitude  \
0      Minnesota             US  2022-07-07  47.871975 -96.095611   
1      Minnesota             US  2022-07-07  44.728364 -94.945537   
2      Minnesota             US  2022-07-07  47.604841 -92.468799   
3      Minnesota             US  2022-07-07  44.022404 -93.226271   
4      Minnesota             US  2022-07-07  45.586130 -96.001637   
5      Minnesota             US  2022-07-07  46.585892 -94.969685   
6      Minnesota             US  2022-07-07  45.038645 -92.884296   
7    Mississippi             US  2022-07-07  34.880845 -88.579962   
8    Mississippi             US  2022-07-07  33.086588 -89.578386   
9    Mississippi             US  2022-07-07  34.816812 -89.189455   

                     location_geom  confirmed  deaths  recovered  active  \
0  POINT(-96.09561104 47.87197525)      964.0    13.0        NaN     NaN   
1  POINT(-94.94553734 44.72836381)     3936.0    58.0        NaN     NaN   
2   POINT(-92.46879855 47.6048407)    52168.0   548.0        NaN     NaN   
3  POINT(-93.22627088 44.02240351)    10974.0    63.0        NaN     NaN   
4  POINT(-96.00163705 45.58613007)     2844.0    14.0        NaN     NaN   
5  POINT(-94.96968506 46.58589171)     4442.0    58.0        NaN     NaN   
6  POINT(-92.88429635 45.03864533)    73267.0   495.0        NaN     NaN   
7  POINT(-88.57996173 34.88084463)    10423.0   161.0        NaN     NaN   
8  POINT(-89.57838564 33.08658803)     5578.0   114.0        NaN     NaN   
9  POINT(-89.18945516 34.81681228)     2353.0    49.0        NaN     NaN   

      fips      admin2               combined_key  
0  27125.0    Red Lake    Red Lake, Minnesota, US  
1  27129.0    Renville    Renville, Minnesota, US  
2  27137.0   St. Louis   St. Louis, Minnesota, US  
3  27147.0      Steele      Steele, Minnesota, US  
4  27149.0     Stevens     Stevens, Minnesota, US  
5  27159.0      Wadena      Wadena, Minnesota, US  
6  27163.0  Washington  Washington, Minnesota, US  
7  28003.0      Alcorn    Alcorn, Mississippi, US  
8  28007.0      Attala    Attala, Mississippi, US  
9  28009.0      Benton    Benton, Mississippi, US

`df.head(10)` digunakan untuk menampilkan 10 indeks teratas dari data frame.

In [30]:
df.tail(10)

province_state country_region        date   latitude  longitude  \
4127526    Puerto Rico             US  2022-02-25  18.059498 -66.613748   
4127527    Puerto Rico             US  2022-02-25  18.010387 -66.256592   
4127528    Puerto Rico             US  2022-02-25  18.147987 -65.976392   
4127529    Puerto Rico             US  2022-02-25  18.431446 -66.213553   
4127530    Puerto Rico             US  2022-02-25  18.271214 -66.702984   
4127531    Puerto Rico             US  2022-02-25  18.409346 -66.337250   
4127532    Puerto Rico             US  2022-02-25  18.428262 -66.397924   
4127533    Puerto Rico             US  2022-02-25  18.070399 -65.896289   
4127534    Puerto Rico             US  2022-02-25  18.080374 -66.858814   
4127535   Rhode Island             US  2022-02-25  41.671063 -71.576416   

                           location_geom  confirmed  deaths  recovered  \
4127526      POINT(-66.613748 18.059498)    16265.0     0.0        NaN   
4127527      POINT(-66.256592 18.010387)     2882.0     0.0        NaN   
4127528      POINT(-65.976392 18.147987)     4926.0     0.0        NaN   
4127529      POINT(-66.213553 18.431446)    11410.0     0.0        NaN   
4127530      POINT(-66.702984 18.271214)     3019.0     0.0        NaN   
4127531       POINT(-66.33725 18.409346)     5513.0     0.0        NaN   
4127532      POINT(-66.397924 18.428262)     9581.0     0.0        NaN   
4127533      POINT(-65.896289 18.070399)     3924.0     0.0        NaN   
4127534      POINT(-66.858814 18.080374)     3660.0     0.0        NaN   
4127535  POINT(-71.57641596 41.67106337)    47000.0   488.0        NaN   

         active     fips       admin2                  combined_key  
4127526     NaN  72113.0        Ponce        Ponce, Puerto Rico, US  
4127527     NaN  72123.0      Salinas      Salinas, Puerto Rico, US  
4127528     NaN  72129.0  San Lorenzo  San Lorenzo, Puerto Rico, US  
4127529     NaN  72137.0     Toa Baja     Toa Baja, Puerto Rico, US  
4127530     NaN  72141.0       Utuado       Utuado, Puerto Rico, US  
4127531     NaN  72143.0    Vega Alta    Vega Alta, Puerto Rico, US  
4127532     NaN  72145.0    Vega Baja    Vega Baja, Puerto Rico, US  
4127533     NaN  72151.0      Yabucoa      Yabucoa, Puerto Rico, US  
4127534     NaN  72153.0        Yauco        Yauco, Puerto Rico, US  
4127535     NaN  44003.0         Kent        Kent, Rhode Island, US

`df.tail(10)` digunakan untuk menampilkan 10 indeks terakhir dari data frame.

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4127536 entries, 0 to 4127535
Data columns (total 13 columns):
 #   Column          Dtype  
---  ------          -----  
 0   province_state  object 
 1   country_region  object 
 2   date            object 
 3   latitude        float64
 4   longitude       float64
 5   location_geom   object 
 6   confirmed       float64
 7   deaths          float64
 8   recovered       float64
 9   active          float64
 10  fips            float64
 11  admin2          object 
 12  combined_key    object 
dtypes: float64(7), object(6)
memory usage: 409.4+ MB


`df.info` digunakan untuk menampilkan informasi nama keseluruhan kolom yang ada pada data frame bersama dengan tipe datanya.

In [32]:
df.describe()

latitude     longitude     confirmed        deaths     recovered  \
count  4.035367e+06  4.035367e+06  4.127517e+06  4.127094e+06  1.435591e+06   
mean   3.591253e+01 -7.206856e+01  7.114377e+04  1.003833e+03  1.636175e+04   
std    1.304303e+01  5.381792e+01  5.635311e+05  6.250453e+03  1.341763e+05   
min   -7.194990e+01 -1.781165e+02 -3.028440e+05 -1.780000e+02 -8.544050e+05   
25%    3.326679e+01 -9.657264e+01  7.590000e+02  1.000000e+01  0.000000e+00   
50%    3.793358e+01 -8.686888e+01  3.573000e+03  5.600000e+01  0.000000e+00   
75%    4.213891e+01 -7.765540e+01  1.544300e+04  2.160000e+02  0.000000e+00   
max    7.200000e+01  1.780650e+02  3.849501e+07  6.246410e+05  6.399531e+06   

             active          fips  
count  1.427512e+06  3.388380e+06  
mean   6.556527e+03  3.228396e+04  
std    9.286123e+04  1.792872e+04  
min   -6.298082e+06  6.000000e+01  
25%    5.500000e+01  1.904100e+04  
50%    4.710000e+02  3.005700e+04  
75%    2.248000e+03  4.702900e+04  
max    1.087776e+07  9.999900e+04

`df.describe()` : menampilkan overview *value* dari dataframe. Yang ditampilkan hanya kolom yang memiliki tipe data float.

In [33]:
a = len(pd.unique(df['province_state']))
print('Total Propinsi : ', a)

Total Propinsi :  760


`len(pd.unique(df['province_state']))` : menampilkan total data nama propinsi yang ada pada kolom province_state tanpa terduplikasi. Terdapat 760 nama propinsi yang ada dalam kolom province_state.

In [34]:
a = len(pd.unique(df['country_region']))
print('Total Negara : ', a)

Total Negara :  253


`len(pd.unique(df['country_region']))` : menampilkan total data nama negara yang ada pada kolom country_region tanpa terduplikasi. Terdapat 253 nama negara yang ada dalam kolom country_region termasuk 1 NaN

In [35]:
pd.isnull(df['province_state']).value_counts()

False    3941731
True      185805
Name: province_state, dtype: int64

syntax `pd.isnull(df['nama_kolom']).value_counts()` digunakan untuk mendeteksi *missing value* dan menghitung total jumlah *missing value* dari kolom yang ditentukan. Nilai True menunjukkan bahwasanya terdapat *missing value* dan nilai False menunjukkan bahwa tidak terdapat *missing value*. Pada contoh diatas merupakan pengecekan *missing value* pada kolom province_state. Terdapat 185.805 *missing value* pada kolom province_state.

In [36]:
a = df['date'].min()
b = df['date'].max()

print('Tanggal Awal  : ', a)
print('Tanggal Akhir : ', b)

Tanggal Awal  :  2020-01-22
Tanggal Akhir :  2023-02-03


Syntax `df['date'].min()` digunakan untuk menampilkan tanggal awal dari kolom 'date'. Sedangkan `df['date'].max()` digunakan untuk menampilkan tanggal terakhir dari kolom 'date'

In [37]:
pd.isna(df['date']).value_counts()

False    4127536
Name: date, dtype: int64

Syntax `pd.isnull(df['date']).value_counts()` digunakan untuk menampilkan total *missing value* pada kolom 'date' yang ternyata tidak ada *missing value* pada kolom 'date' dibuktikan dengan tidak adanya nilai True.

In [38]:
a = len(pd.unique(df['date']))
print('Total Hari : ', a)

Total Hari :  1109


Syntax `a = len(pd.unique(df['date']))` digunakan untuk menghitung jumlah total hari. Toral hari yang dimaksud adalah range antara tanggal awal sampai tanggal akhir. Alasan menggunakan `a = len(pd.unique(df['date']))` adalah karena tanggal pada kolom date tidak terdapat *missing value* dibuktikan dengan pengecekan menggunakan `pd.isnull(df['date']).value_counts()` pada baris kode sebelumnya sehingga dapat dihitung data unik yang menghasikan 1109 data yang merupakan representasi total hari dari awal sampai akhir data.

## Data Cleaning

### Merubah Nama Kolom

Merubah nama kolom agar dalam proses pengolahan data menjadi lebih jelas dan lebih mengerti kolom yang dimaksud. Kolom yang akan diubah namnaya adalah kolom `admin2` karena data di dalam kolom `admin2` merupakan nama kota. Kolom `admin2` akan diubah namanya menjadi kolom `city_name`. Selain nama kolom admin2 yang akan diubah, kolom `location_geom` juga akan diubah karena data yang ditampilkan di kolom tersebut adalah koordinat, maka nama kolom akan diubah menjadi `coordinates`.

In [39]:
df

province_state country_region        date   latitude  longitude  \
0            Minnesota             US  2022-07-07  47.871975 -96.095611   
1            Minnesota             US  2022-07-07  44.728364 -94.945537   
2            Minnesota             US  2022-07-07  47.604841 -92.468799   
3            Minnesota             US  2022-07-07  44.022404 -93.226271   
4            Minnesota             US  2022-07-07  45.586130 -96.001637   
...                ...            ...         ...        ...        ...   
4127531    Puerto Rico             US  2022-02-25  18.409346 -66.337250   
4127532    Puerto Rico             US  2022-02-25  18.428262 -66.397924   
4127533    Puerto Rico             US  2022-02-25  18.070399 -65.896289   
4127534    Puerto Rico             US  2022-02-25  18.080374 -66.858814   
4127535   Rhode Island             US  2022-02-25  41.671063 -71.576416   

                           location_geom  confirmed  deaths  recovered  \
0        POINT(-96.09561104 47.87197525)      964.0    13.0        NaN   
1        POINT(-94.94553734 44.72836381)     3936.0    58.0        NaN   
2         POINT(-92.46879855 47.6048407)    52168.0   548.0        NaN   
3        POINT(-93.22627088 44.02240351)    10974.0    63.0        NaN   
4        POINT(-96.00163705 45.58613007)     2844.0    14.0        NaN   
...                                  ...        ...     ...        ...   
4127531       POINT(-66.33725 18.409346)     5513.0     0.0        NaN   
4127532      POINT(-66.397924 18.428262)     9581.0     0.0        NaN   
4127533      POINT(-65.896289 18.070399)     3924.0     0.0        NaN   
4127534      POINT(-66.858814 18.080374)     3660.0     0.0        NaN   
4127535  POINT(-71.57641596 41.67106337)    47000.0   488.0        NaN   

         active     fips     admin2                combined_key  
0           NaN  27125.0   Red Lake     Red Lake, Minnesota, US  
1           NaN  27129.0   Renville     Renville, Minnesota, US  
2           NaN  27137.0  St. Louis    St. Louis, Minnesota, US  
3           NaN  27147.0     Steele       Steele, Minnesota, US  
4           NaN  27149.0    Stevens      Stevens, Minnesota, US  
...         ...      ...        ...                         ...  
4127531     NaN  72143.0  Vega Alta  Vega Alta, Puerto Rico, US  
4127532     NaN  72145.0  Vega Baja  Vega Baja, Puerto Rico, US  
4127533     NaN  72151.0    Yabucoa    Yabucoa, Puerto Rico, US  
4127534     NaN  72153.0      Yauco      Yauco, Puerto Rico, US  
4127535     NaN  44003.0       Kent      Kent, Rhode Island, US  

[4127536 rows x 13 columns]

`df.head()` digunakan untuk menampilkan tabel sebelum kolom `admin2` diubah namanya menjadi `city_name` dan kolom `location_geom` menjadi `coordinates`

In [40]:
df.rename({'combined_key':'Location', 'admin2':'city_name', 'location_geom':'coordinates'}, axis=1, inplace=True)

`df.rename({'admin2':'city_name', 'location_geom':'coordinates'}, axis=1, inplace=True)` digunakan untuk merubah nama kolom `admin2` menjadi `city_name` dan `location_geom` menjadi `coordinates` dan `combined_key` menjadi `Location`

In [41]:
df.head()

province_state country_region        date   latitude  longitude  \
0      Minnesota             US  2022-07-07  47.871975 -96.095611   
1      Minnesota             US  2022-07-07  44.728364 -94.945537   
2      Minnesota             US  2022-07-07  47.604841 -92.468799   
3      Minnesota             US  2022-07-07  44.022404 -93.226271   
4      Minnesota             US  2022-07-07  45.586130 -96.001637   

                       coordinates  confirmed  deaths  recovered  active  \
0  POINT(-96.09561104 47.87197525)      964.0    13.0        NaN     NaN   
1  POINT(-94.94553734 44.72836381)     3936.0    58.0        NaN     NaN   
2   POINT(-92.46879855 47.6048407)    52168.0   548.0        NaN     NaN   
3  POINT(-93.22627088 44.02240351)    10974.0    63.0        NaN     NaN   
4  POINT(-96.00163705 45.58613007)     2844.0    14.0        NaN     NaN   

      fips  city_name                  Location  
0  27125.0   Red Lake   Red Lake, Minnesota, US  
1  27129.0   Renville   Renville, Minnesota, US  
2  27137.0  St. Louis  St. Louis, Minnesota, US  
3  27147.0     Steele     Steele, Minnesota, US  
4  27149.0    Stevens    Stevens, Minnesota, US

`df.head()` digunakan untuk menampilkan head tabel guna menunjukkan perubahan nama kolom `admin2` menjadi `city_name` dan kolom `location_geom` menjadi `coordinates`

### Menghapus Kolom yang Tidak Diperlukan

 

*   Kolom `latitude` dan `longitude` dihapus karena `coordinates` mewakili nilai yang ada pada `latitude` dan `longitude`.
*   Kolom `fips` dihapus karena `fips` hanya digunakan di US. Negara selain US tidak memiliki `fips`.



In [42]:
df.isna().sum()

province_state     185805
country_region          1
date                    0
latitude            92169
longitude           92169
coordinates         92169
confirmed              19
deaths                442
recovered         2691945
active            2700024
fips               739156
city_name          734209
Location             7618
dtype: int64

In [43]:
df.drop('latitude', axis=1, inplace=True) #sudah ada di coordinates

In [44]:
df.drop('longitude', axis=1, inplace=True) #sudah ada di coordinates

In [45]:
df.drop('fips', axis=1, inplace=True) #hanya ada di Us, selain US fips bernilai Null

`df.drop('nama_kolom', axis=1, inplace=True)` merupakan syntax untuk menghapus kolom.

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4127536 entries, 0 to 4127535
Data columns (total 10 columns):
 #   Column          Dtype  
---  ------          -----  
 0   province_state  object 
 1   country_region  object 
 2   date            object 
 3   coordinates     object 
 4   confirmed       float64
 5   deaths          float64
 6   recovered       float64
 7   active          float64
 8   city_name       object 
 9   Location        object 
dtypes: float64(4), object(6)
memory usage: 314.9+ MB


`df` digunakan untuk menapilkan dataframe. Pada tampilan diatas sudah terlihat kolom longitude dan latitude sudah terhapus. Akan dibuktikan juga dengan df.shape agar terlihat bahwa 2 kolom sudah terhapus.

In [47]:
df.shape

(4127536, 10)

In [48]:
df.isna().sum()

province_state     185805
country_region          1
date                    0
coordinates         92169
confirmed              19
deaths                442
recovered         2691945
active            2700024
city_name          734209
Location             7618
dtype: int64

In [49]:
df.dropna(subset=['country_region'], inplace = True) #hapus 1 NaN pada coutry_region

In [50]:
df.isna().sum() #menampilkan 1 missing value pada country_region yang telah dihapus

province_state     185804
country_region          0
date                    0
coordinates         92168
confirmed              19
deaths                442
recovered         2691945
active            2700024
city_name          734208
Location             7617
dtype: int64

In [51]:
df.dropna(subset=['coordinates'], inplace = True) #hapus nan di coordinates

In [52]:
df.dropna(subset=['active'], inplace = True) #hapus nan di active

In [54]:
df.shape # Setelah melakukan dropna total row pada dataframe tinggal 1.398.950 dan 10 column setelah melakukan drop pada kolom longitude, latitude dan fips.

(1398950, 10)

### Mengisi nilai null dan NaN pada `city_name` dan `province_state`

In [55]:
df.isna().sum() # Menampilkan total NaN pada setiap kolom yang ada.

province_state     84977
country_region         0
date                   0
coordinates            0
confirmed              0
deaths                 0
recovered              0
active                 0
city_name         313622
Location               0
dtype: int64

In [56]:
df['province_state'].fillna('Province', inplace=True) # Manipulasi untuk mengisi string yang kosong pada kolom province_state dengan isi 'Province'

In [57]:
df['city_name'].fillna('City', inplace=True) # Manipulasi untuk mengisi string yang kosong pada kolom city_name dengan isi 'City'

In [58]:
df.isna().sum() # Memanggil fungsi isna() untuk pengecekan nilai null/Nan pada setiap kolom yang ada dan menampilkan jumlah keseluruhan.

province_state    0
country_region    0
date              0
coordinates       0
confirmed         0
deaths            0
recovered         0
active            0
city_name         0
Location          0
dtype: int64

Dari hasil output diatas dapat  dilihat sudah tidak terdapat nilai null maupun NaN pada setiap kolom yang ada. Data sudah siap untuk dieksplorasi.

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1398950 entries, 648 to 4125224
Data columns (total 10 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   province_state  1398950 non-null  object 
 1   country_region  1398950 non-null  object 
 2   date            1398950 non-null  object 
 3   coordinates     1398950 non-null  object 
 4   confirmed       1398950 non-null  float64
 5   deaths          1398950 non-null  float64
 6   recovered       1398950 non-null  float64
 7   active          1398950 non-null  float64
 8   city_name       1398950 non-null  object 
 9   Location        1398950 non-null  object 
dtypes: float64(4), object(6)
memory usage: 117.4+ MB


In [60]:
df['active']=df['active'].astype(int)
df['deaths']=df['deaths'].astype(int)
df['recovered']=df['recovered'].astype(int)
df['confirmed']=df['confirmed'].astype(int)
df['date']=pd.to_datetime(df['date'])

Merubah tipe data `active` `deaths` `recovered` `confirmed` dari float menjadi tipe data integer dan `'date'` dari string menjadi tipe data date

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1398950 entries, 648 to 4125224
Data columns (total 10 columns):
 #   Column          Non-Null Count    Dtype         
---  ------          --------------    -----         
 0   province_state  1398950 non-null  object        
 1   country_region  1398950 non-null  object        
 2   date            1398950 non-null  datetime64[ns]
 3   coordinates     1398950 non-null  object        
 4   confirmed       1398950 non-null  int64         
 5   deaths          1398950 non-null  int64         
 6   recovered       1398950 non-null  int64         
 7   active          1398950 non-null  int64         
 8   city_name       1398950 non-null  object        
 9   Location        1398950 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(5)
memory usage: 117.4+ MB


## Data Exploration